persistent/Database/Persist/Sql/Class.hs

```diff
@@ -62,23 +63,28 @@ instance (PersistEntity a, PersistEntityBackend a ~ SqlBackend) => RawSql (Key a
   rawSqlProcessRow         = keyFromValues
 
 instance (PersistEntity a, PersistEntityBackend a ~ SqlBackend) => RawSql (Entity a) where
-    rawSqlCols escape = ((+1) . length . entityFields &&& process) . entityDef . Just . entityVal
+    rawSqlCols escape ent = (length sqlFields, [intercalate ", " sqlFields])
         where
-          process ed = (:[]) $
-                       intercalate ", " $
-                       map ((name ed <>) . escape) $
-                       (fieldDB (entityId ed) :) $
-                       map fieldDB $
-                       entityFields ed
-          name ed = escape (entityDB ed) <> "."
-
+          sqlFields = map (((name <> ".") <>) . escape)
+              $ map fieldDB
+              $ keyFields eDef ++ entityFields eDef
+          name = escape (entityDB eDef)
+          eDef = entityDef $ Just (entityVal ent)
     rawSqlColCountReason a =
         case fst (rawSqlCols (error "RawSql") a) of
           1 -> "one column for an 'Entity' data type without fields"
           n -> show n ++ " columns for an 'Entity' data type"
-    rawSqlProcessRow (idCol:ent) = Entity <$> fromPersistValue idCol
-                                          <*> fromPersistValues ent
-    rawSqlProcessRow _ = Left "RawSql (Entity a): wrong number of columns."
+    rawSqlProcessRow =
+        let nKeyFields = length $ keyFields $ entityDef (Nothing :: Maybe (Entity a))
+            processRow row = case splitAt nKeyFields row of
+                (rowKey, rowVal) -> Entity <$> keyFromValues rowKey
+                                           <*> fromPersistValues rowVal
+        in processRow
+
+keyFields :: EntityDef -> [FieldDef]
+keyFields ed = case entityPrimary ent of
+    Nothing   -> [entityId ed]
+    Just pdef -> compositeFields pdef
```

Comment:
Should this go in `Database.Persist.Sql.Util`, which seems to have all the other composite key helper functions?